Subject: [PATCH] kdump: clear and EOI IPI for kexec CPU
From: Haren Myneni <haren@us.ibm.com>
To: Paul Mackerras <paulus@samba.org>
Cc: linuxppc-dev@ozlabs.org, ellerman@au1.ibm.com, Milton Miller <miltonm@bga.com>, Olaf Hering <olh@suse.de>
Date: Wed, 22 Mar 2006 21:18:24 -0800

Bug 165758 - LTC23173-kexec CPU is not responding to an IPI

Some times, the kexec CPU is not reponding to an IPI during kdump boot.
It is causing the system in xmon. Noticed on power-4 (non-lpar). This
patch clears and EOI IPI for kexec CPU as well before the kdump boot
started.

Signed-off-by: Haren Myneni <haren@us.ibm.com>
Signed-off-by: Olaf Hering <olh@suse.de>

---
 arch/powerpc/platforms/pseries/xics.c |   24 ++++++++++++++----------
 1 file changed, 14 insertions(+), 10 deletions(-)

Index: linux-2.6.16-olh-kdump/arch/powerpc/platforms/pseries/xics.c
===================================================================
--- linux-2.6.16-olh-kdump.orig/arch/powerpc/platforms/pseries/xics.c
+++ linux-2.6.16-olh-kdump/arch/powerpc/platforms/pseries/xics.c
@@ -641,22 +641,26 @@ void xics_teardown_cpu(int secondary)
 	iosync();
 
 	/*
+	 * Clear IPI
+	 */
+	ops->qirr_info(cpu, 0xff);
+
+	/*
+	 * we need to EOI the IPI if we got here from kexec down IPI
+	 *
+	 * probably need to check all the other interrupts too
+	 * should we be flagging idle loop instead?
+	 * or creating some task to be scheduled?
+	 */
+	ops->xirr_info_set(cpu, XICS_IPI);
+	/*
 	 * Some machines need to have at least one cpu in the GIQ,
 	 * so leave the master cpu in the group.
 	 */
-	if (secondary) {
-		/*
-		 * we need to EOI the IPI if we got here from kexec down IPI
-		 *
-		 * probably need to check all the other interrupts too
-		 * should we be flagging idle loop instead?
-		 * or creating some task to be scheduled?
-		 */
-		ops->xirr_info_set(cpu, XICS_IPI);
+	if (secondary)
 		rtas_set_indicator(GLOBAL_INTERRUPT_QUEUE,
 			(1UL << interrupt_server_size) - 1 -
 			default_distrib_server, 0);
-	}
 }
 
 #ifdef CONFIG_HOTPLUG_CPU
